### PR TITLE
bug/brapi studies fixes additions

### PR DESCRIPTION
**Pull request by ctucker3**
Description 
-----------------------------------------------------
- Added generic brapi endpoint implementation for required fields
- Added check to brapi endpoints for proper json body
- Return 404 if trialDbId is not found in db
- Return 404 if locationDbId is not found in db
- Remove requirement for season, experimentalDesign,studyDescription
- Return 409 if a study (or program, trial) with that name already exists
- Add additional info to studies
- Add miscellaneous study type (is treated as phenotyping trial in backend)
  - If a study type is passed that doesn't match an existing type, is stores it as miscellaneous and save the type value. 
- Require `locationDbId` for POST studies to make it consistent with PUT studies (was using locationName before)

Special notes: 
- Not having a design type will cause a pop up modal error in the front end that is closable. The error doesn't affect the other information from loading. If you click on the 'Field Layout Tools and Phenotype Heat Map', an error will occur and the loading modal will get stuck on the page. 
- Needed changes made in the trials PR that is open, so is based off from that: https://github.com/solgenomics/sgn/pull/3533

Related Issue: https://github.com/solgenomics/sgn/issues/3527

Checklist 
---------------------------------------------------------------------------------------
- [ ] Refactoring only
- [ ] Documentation only
- [ ] Fixture update only
- [x ] Bug fix
  - [ ] The relevant issue has been closed.
  - [ ] Further work is required.
- [ ] New feature
  - [ ] Relevant tests have been created and run.
  - [ ] Data was added to the fixture
    - [x ] Data was added via a patch in `/t/data/fixture/patches/`.
  - [ ] User-Facing Change
    - [ ] The user manual in `/docs` has been updated.
  - [ x] Any new Perl has been documented using **perldoc**.
  - [ ] Any new JavaScript has been documented using **JSDoc**.
  - [ ] Any new _legacy_ JavaScript has been moved from `/js` to `/js/source/legacy`.

### CODE REVIEW

File: lib/CXGN/Trial/TrialCreate.pm [hunk 1]
```diff
@@ -146,13 +146,14 @@ has 'dbh' => (is  => 'rw',predicate => 'has_dbh', required => 1,);
...
-has 'design' => (isa => 'HashRef[HashRef]|Undef', is => 'rw', predicate => 'has_design', required => 1);
+has 'trial_year' => (isa => 'Maybe[Str]', is => 'rw', predicate => 'has_trial_year', required => 0,);
+has 'trial_description' => (isa => 'Maybe[Str]', is => 'rw', predicate => 'has_trial_description', required => 0,);
+has 'trial_location' => (isa => 'Maybe[Str]', is => 'rw', predicate => 'has_trial_location', required => 0,);
+has 'design_type' => (isa => 'Maybe[Str]', is => 'rw', predicate => 'has_design_type', required => 0);
+has 'design' => (isa => 'HashRef[HashRef]|Undef', is => 'rw', predicate => 'has_design', required => 0);
 has 'trial_name' => (isa => 'Str', is => 'rw', predicate => 'has_trial_name', required => 1);
-has 'trial_type' => (isa => 'Str', is => 'rw', predicate => 'has_trial_type', required => 0);
+has 'trial_type' => (isa => 'Str', is => 'rw', predicate => 'has_trial_type', required => 1);
+has 'trial_type_value' => (isa => 'Str', is => 'rw', predicate => 'has_trial_type', required => 0);
```
Comment (nickmorales) [inline lib/CXGN/Trial/TrialCreate.pm:156]: This does not compile because there are now 2 has_trial_type predicates

Comment (ctucker3) [reply lib/CXGN/Trial/TrialCreate.pm:156]: Good catch! I'll fix that today. I am also going to remove the optional design code and make it required again. It seems too central to other BreedBase features to not have a design code.

Comment (ctucker3) [reply lib/CXGN/Trial/TrialCreate.pm]: Predicate has been fixed. I made design type required again (`experimentalDesign.PUI` in brapi).

File: db/00139/AddMiscellaneousTrialType.pm [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        my $new_cvterm = $schema->resultset("Cv::Cvterm")->create_with(
+            {
+                name => 'misc_trial',
+                definition => 'A trial type that holds any trial type',
+                cv   => 'project_type',
+            });
+    };
+
+    try {
+        $schema->txn_do($coderef);
```
Comment (nickmorales) [inline db/00139/AddMiscellaneousTrialType.pm:74]: $schema is not defined and so gives an error/crashes

Comment (ctucker3) [reply db/00139/AddMiscellaneousTrialType.pm:73]: fixed and modified a test so that this new type would be used in the tests.

**Comment by ctucker3** [comment]
This one is good to go, depending on approval, now that the brapi trials PR is merge in.

**Review by MFlores2021 — APPROVED** [review]

**Review by lukasmueller — APPROVED** [review]
:-)

**Pull request merged**